src/lib/gadgets/rot.ts

```diff
@@ -0,0 +1,118 @@
+import { Field } from '../field.js';
+import { Provable } from '../provable.js';
+import { Fp } from '../../bindings/crypto/finite_field.js';
+import * as Gates from '../gates.js';
+
+export { rot, rotate };
+
+const MAX_BITS = 64 as const;
+
+function rot(field: Field, bits: number, direction: 'left' | 'right' = 'left') {
+  // Check that the rotation bits are in range
+  if (bits < 0 || bits > MAX_BITS) {
+    throw Error(`rot: expected bits to be between 0 and 64, got ${bits}`);
+  }
+
+  if (direction !== 'left' && direction !== 'right') {
+    throw Error(
+      `rot: expected direction to be 'left' or 'right', got ${direction}`
+    );
+  }
+
+  if (field.isConstant()) {
+    checkMaxBits(field);
+    return new Field(Fp.rot(field.toBigInt(), bits, direction));
+  }
+  const [rotated] = rotate(field, bits, direction);
+  return rotated;
+}
+
+function rotate(
+  field: Field,
+  bits: number,
+  direction: 'left' | 'right' = 'left'
+): [Field, Field, Field] {
+  // Check as the prover, that the input is at most 64 bits.
+  Provable.asProver(() => {
+    checkMaxBits(field);
+  });
```

Comment:
I think this check should be removed, reasoning here: https://github.com/o1-labs/o1js/pull/1177#discussion_r1362119516

Reply:
Fixed in https://github.com/o1-labs/o1js/pull/1182/commits/36415ad666933783b8a3596dd196c75b31b68c69